cardano-cli/src/Cardano/CLI/Shelley/Run.hs

```diff
@@ -36,19 +36,36 @@ data ShelleyClientCmdError
   | ShelleyCmdKeyError !ShelleyKeyCmdError
   deriving Show
 
-renderShelleyClientCmdError :: ShelleyClientCmdError -> Text
-renderShelleyClientCmdError err =
-  case err of
-    ShelleyCmdAddressError addrCmdErr -> renderShelleyAddressCmdError addrCmdErr
-    ShelleyCmdGenesisError genesisCmdErr -> renderShelleyGenesisCmdError genesisCmdErr
-    ShelleyCmdGovernanceError govCmdErr -> renderShelleyGovernanceError govCmdErr
-    ShelleyCmdNodeError nodeCmdErr -> renderShelleyNodeCmdError nodeCmdErr
-    ShelleyCmdPoolError poolCmdErr -> renderShelleyPoolCmdError poolCmdErr
-    ShelleyCmdStakeAddressError stakeAddrCmdErr -> renderShelleyStakeAddressCmdError stakeAddrCmdErr
-    ShelleyCmdTextViewError txtViewErr -> renderShelleyTextViewFileError txtViewErr
-    ShelleyCmdTransactionError txErr -> renderShelleyTxCmdError txErr
-    ShelleyCmdQueryError queryErr -> renderShelleyQueryCmdError queryErr
-    ShelleyCmdKeyError keyErr -> renderShelleyKeyCmdError keyErr
+-- Identity monad is used to avoid boilerplate
+renderShelleyClientCmdError :: ShelleyCommand -> ShelleyClientCmdError -> Identity Text
+renderShelleyClientCmdError cmd err = do
+  cmdError <- case err of
+                ShelleyCmdAddressError addrCmdErr ->
+                  return $ renderShelleyAddressCmdError addrCmdErr
+                ShelleyCmdGenesisError genesisCmdErr ->
+                  return $ renderShelleyGenesisCmdError genesisCmdErr
+                ShelleyCmdGovernanceError govCmdErr ->
+                  return $ renderShelleyGovernanceError govCmdErr
+                ShelleyCmdNodeError nodeCmdErr ->
+                  return $ renderShelleyNodeCmdError nodeCmdErr
+                ShelleyCmdPoolError poolCmdErr ->
+                  return $ renderShelleyPoolCmdError poolCmdErr
+                ShelleyCmdStakeAddressError stakeAddrCmdErr ->
+                  return $ renderShelleyStakeAddressCmdError stakeAddrCmdErr
+                ShelleyCmdTextViewError txtViewErr ->
+                  return $ renderShelleyTextViewFileError txtViewErr
+                ShelleyCmdTransactionError txErr ->
+                  return $ renderShelleyTxCmdError txErr
+                ShelleyCmdQueryError queryErr ->
+                  return $ renderShelleyQueryCmdError queryErr
+                ShelleyCmdKeyError keyErr ->
+                  return $ renderShelleyKeyCmdError keyErr
+
+  return $ "Shelley command failed: "
+         <> renderShelleyCommand cmd
+         <> "  Error: "
+         <> cmdError
```

Comment:
If it's for a better way to implement the `renderShelleyClientCmdError` function, the shocking thing is you don't need a monad to use the `do` notation.  For example this works with no monad in sight:

```
moo :: Int -> String
moo n = do
  x <- case n of
    1 -> "1"
    2 -> "2"
  return (x <> "!")
```

Reply:
Isn't `String` = `[Char]` so we would be in the list monad?

Reply:
What is wrong with
```
mconcat [ "Shelley command failed: ". renderShelleyCommand cmd, "  Error: ",  cmdError ]
``` 
?

Reply:
Nothing wrong.  Just more of a curiosity.

Reply:
> Isn't `String` = `[Char]` so we would be in the list monad? I've also re-pushed some changes to get rid of the `Identity` monad. Would probably be easier to look at the diff and see my rational for using `Identity`

I made a typo in the origin code where I used `<-` and `return`.  If you remove all the monad things, `do` still works.  I don't actually write code non-monadic code with a `do` block, and not suggesting it.  😁